### PR TITLE
Cadence v5 feature upgrade

### PR DESCRIPTION
**Pull request by za419**
This PR is currently incomplete and untested, but I don't mind putting it out for comments in this state.

I also realize I probably should have broken up the scope here and merged each feature individually - Oh well.

### CODE REVIEW

File: bot.js [hunk 1]
```diff
@@ -117,13 +117,22 @@ const stream = bot.voice.createBroadcast();
 // This variable will track our current stream status, for reporting reasons
 let streamStatus = "Startup";
 
+// This will be the path we connect to.
+// By default, we shall use fallback paths set in config
+// But, if configured to try to auto-fetch it, we'll grab the new path and put it here.
```
Comment (za419) [inline bot.js:184]: I considered also adding behavior for the fallback path to get used later if the stream fails to connect or dies - As a hardcoded redundancy, not just a block to the listenURL not getting updated.

I spent about 10 seconds on it and didn't think of a great way to differentiate a temporary network issue from a condition where we should set back to fallback, except implementing a reconnect counter and tuning it for some big number, which seems sad.

Thoughts? Would this be expected behavior I should put in?

File: bot.js [hunk 2]
```diff
@@ -138,6 +147,38 @@ function beginGlobalPlayback() {
     }
 }
 
+// If the stream path should be set automatically, try to override our streamPath from the upstream.
```
Comment (za419) [inline bot.js:213]: With this, the minimal configuration for a new CadenceBot deployment connected to a new CadenceBot server would be something like

```json
{
  "API": {
    "aria": {
      "prefix": "https://NEW_SERVER_HERE"
    }
  }
}
```

as long as default command triggers and such are comfortable for the new IA.

I don't think we'll get it any lower than that, so @kenellorando if you want to anticipate the install script generating a `config.json` like that for CadenceBot feel free to do so.

File: bot.js [hunk 3]
```diff
@@ -1169,22 +1208,26 @@ function command(message) {
                     );
 
                     // Grab the report of how much time is left from the response, and parse it into a string
-					try {
```
Comment (za419) [inline bot.js:1172]: Woah, woah, woah - How the heck did *tabs* get in here!?

Goddammit... Um, me? VSCode?

Yeah, let's go with VSCode.

File: bot.js [hunk 4]
```diff
@@ -1327,6 +1370,52 @@ function command(message) {
...
+        log.info("Making a request to " + url);
+        request.get({ url, form: {} }, (err, response, body) => {
+            if (!err && body != null) {
+                log.info("Request succeeded.");
+                log.debug("Received body:");
+                log.debug(body);
+                let history;
+                if ((body = "" || (history = JSON.parse(body)).length === 0)) {
+                    log.info("Results are empty. Replying appropriately.");
+                    message.reply("Cadence has no song history at the moment.");
```
Comment (za419) [inline bot.js:1474]: @kenellorando feel free to commend on this response and the others for `history`

File: default-config.json [hunk 5]
```diff
@@ -183,13 +187,17 @@
...
       "nowplaying": "/api/nowplaying/metadata",
+      "history": "/api/history",
+      "listenurl": "/api/listenurl",
       "key": false
     },
     "stream": {
-      "prefix": "https://stream.cadenceradio.com",
-      "stream": "/cadence1"
+      "automatic": true,
+      "protocol": "https",
```
Comment (za419) [inline default-config.json:199]: The Cadence API reference says we choose this - I decided to enforce that fallback and primary both use the same protocol, but not to enforce that that protocol matches the main API prefix.

I'm open to disagreement with that choice though.

File: bot.js (hunk 4 above)
Comment (kenellorando) [reply bot.js:1474]: Looks fine. Cadence's UI just says `No history available (yet).`

https://github.com/kenellorando/cadence/blob/362800041145694dbf61da01833253078b670941/cadence/server/public/js/aria.js#L96-L97

**Review by kenellorando — APPROVED** [review]

**Pull request merged**